Add templates for Web Extensions page types (#42204)
mdn/content: files/en-us/mdn/writing_guidelines/page_structures/page_types/webextension_api_event_page_template/index.md

---
title: WebExtension API event page template
slug: MDN/Writing_guidelines/Page_structures/Page_types/WebExtension_API_event_page_template
page-type: mdn-writing-guide
sidebar: mdnsidebar
---

> This page describes the structure and content of a WebExtension API event page. To get started quickly, copy the template file at [`files/templates/webextensions/api-event.md`](https://github.com/mdn/content/blob/main/files/templates/webextensions/api-event.md) in the content repository. Save it as `index.md` in a directory with a lowercase name matching the slug. Then use this guidance to fill in each section.

Reviewer: As the template includes the preferred wording, can we be more explicit about its use, e.g.,
```suggestion
This page describes the structure and content of a WebExtension API event page. Use this guide in conjunction with the template file at [`files/templates/webextensions/api-event.md`](https://github.com/mdn/content/blob/main/files/templates/webextensions/api-event.md) in the content repository. To create a page, save the template as `index.md` in a directory with a lowercase name matching the slug. Then follow this guidance to fill in each section.
```

## Page front matter

> The front matter at the start of the page defines page metadata. The values must be updated appropriately for the WebExtension API event.

Reviewer: ```suggestion
The front matter at the start of the page defines page metadata. Update the values appropriately for the WebExtension API event.
```

```md
---
title: Namespace.onEventName
slug: Mozilla/Add-ons/WebExtensions/API/Namespace/onEventName
page-type: webextension-api-event
status:
  - deprecated
  - experimental
browser-compat: webextensions.api.Namespace.onEventName
sidebar: addonsidebar
---
```

- **title**
  - : Title heading displayed at the start of the page.
    Format as `"Namespace.onEventName"`. For example, the [`runtime.onMessage`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage) event page has a _title_ of `runtime.onMessage`.
- **slug**
  - : The end of the URL path after `https://developer.mozilla.org/en-US/docs/`.

>     This is formatted like `Mozilla/Add-ons/WebExtensions/API/Namespace/onEventName`.

Reviewer: As we included an example for title, I think we should continue with that where applicable using the same event.
```suggestion
    This is formatted like `Mozilla/Add-ons/WebExtensions/API/Namespace/onEventName`. For example, the [`runtime.onMessage`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage) event page's slug is `Mozilla/Add-ons/WebExtensions/API/runtime/onMessage`.
```

- **page-type**
  - : Always `webextension-api-event`.
- **status**
  - : Include status flags as appropriate: `experimental`, `deprecated`, or `non-standard`. You may also need to add banner content to the page, see "Top-of-page macros". Remove the `status` field if no flags apply.

    > [!NOTE]
    > Unlike web features, WebExtension API page status blocks aren't automatically updated based on the browser compatibility data.

- **browser-compat**

>   - : Replace `Namespace` and `onEventName` in the placeholder `webextensions.api.Namespace.onEventName` with the appropriate values used in the [browser-compat-data repo](https://github.com/mdn/browser-compat-data). For guidance on creating this data, see [Browser Compatibility tables and Browser Compatibility Data (BCD)](/en-US/docs/MDN/Writing_guidelines/Page_structures/Compatibility_tables). The toolchain uses the key to populate the compatibility section (replacing the `{{Compat}}` macro).

Reviewer: ```suggestion
  - : Replace `Namespace` and `onEventName` in the placeholder `webextensions.api.Namespace.onEventName` with the values used in the [browser-compat-data repo](https://github.com/mdn/browser-compat-data). For example, for the [`runtime.onMessage`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage) event, the value is `webextensions.api.runtime.onMessage`.

  For guidance on creating this data, see [Browser Compatibility tables and Browser Compatibility Data (BCD)](/en-US/docs/MDN/Writing_guidelines/Page_structures/Compatibility_tables). The toolchain uses the key to populate the compatibility section (replacing the `{{Compat}}` macro).
```

- **sidebar**
  - : Always set to `addonsidebar`. The sidebar content is determined by the directory structure and front matter. It's therefore important to place the page in the correct location under `Mozilla/Add-ons/WebExtensions/API/`.

## Top-of-page macros

> You should add certain macro calls to the content section (immediately following the page front matter). For WebExtension pages, you must add these macros as needed:

Reviewer: ```suggestion
You should add certain macro calls to the content section (immediately following the page front matter). For WebExtension pages, add these macros as needed:
```

- `{{SeeCompatTable}}` — add this if the event is [experimental](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#experimental). Also add `experimental` to the `status` list in the front matter.
- `{{Deprecated_Header}}` — add this if the event is [deprecated](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#deprecated). Also add `deprecated` to the `status` list in the front matter. In general, WebExtension pages prefer a custom deprecation warning with information about the replacement feature, for example:

  ```md
  > [!WARNING]
  > This event is deprecated. Use \{{WebExtAPIRef("Namespace.onReplacementEvent")}} instead.
  ```

Remove any macros or banners that do not apply.

> ## Introductory paragraph

Reviewer: In general, we're providing a comprehensive description section. As noted in further content feedback, I'm not sure it makes sense to move detailed descriptive information to the end of the page.

> Begin the content on the page with an introductory paragraph describing when the event fires and why it is useful. This should ideally be one or two short sentences. There is no need to repeat the name and namespace since the title already communicates this.

Reviewer: ```suggestion
Begin the page content with an introductory paragraph that explains when the event fires and why it is useful. This should ideally be one or two short sentences. There is no need to repeat the name and namespace, as the title already communicates this.
```

## Syntax section

Show how to register a listener for the event. Include the three standard event functions (`addListener`, `removeListener`, `hasListener`) with brief descriptions. If the event supports filters or extra parameters for `addListener` (for example, some `webRequest` events accept a `filter` and `extraInfoSpec`), show that pattern instead of or in addition to the basic form.

## addListener syntax section

### Parameters subsection

> Describe the arguments passed to the listener callback function. Use a definition list with one term and definition per parameter. If the listener receives a simple object, you can document its fields inline. For complex objects with many fields, briefly reference them here and document the full shape in a separate `## Additional objects` section (see below).

Reviewer: However, also see my comments under the additional object section.
```suggestion
Describe the parameters passed to the add listener function using a [definition list](/en-US/docs/MDN/Writing_guidelines/Howto/Markdown_in_MDN#definition_lists). If the listener receives a simple object, document its arguments inline. For complex objects, briefly describe the objects and document their full shape in the [`Additional objects`](#additional_objects) section.
```

> ### Return value subsection

Reviewer: As none of the events that return a value use a "Return value" section, should we mock up how it would be used to confirm usefulness and position? e.g., should it be part of the listener documentation, as it appears to be most commonly at the moment. If we do adopt it, we should include a form of words in the template to ensure standard presentation of the content.

> If the listener is expected to return a value or a `Promise` (for example, to block or modify a request in `webRequest` events), document the expected return value here. If the listener does not return anything, you can omit this subsection.

Reviewer: As events don't necessarily return value promised by default, should the section also include a description of the circumstances under which one is returned?
```suggestion
If the listener can return a value or a `Promise` (for example, to block or modify a request in `webRequest` events), document the expected return value here. If the listener does not return anything, omit this subsection.
```

## Additional objects section

> When the listener callback receives complex objects with many fields (for example, a `details` object), document the full shape of each object in a separate `## Additional objects` section placed between `addListener syntax` and `Examples`. Each object gets an H3 (`###`) heading, and its fields are listed in a definition list. This pattern is used widely across existing event pages — see [`tabs.onActivated`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onActivated) for a simple example and [`webRequest.onBeforeSendHeaders`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeSendHeaders) for a complex one.

Reviewer: I think we need to either provide specific guidance on what constitutes an object with "many fields" or treat as an exception. For example, on [tabs.onActivated](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onActivated), the use of an additional field section seems fairly pointless (it's documents one object with 3 fields). Even the section naming seems a bit odd, given that in many cases these objects aren't "additional", it's more "listener object details".

It may also be appropriate to consider whether this pattern is necessary For example, in documentation such as that for [Stripe](https://docs.stripe.com/api/customers/object#customer_object-address) or [commercetools](https://docs.commercetools.com/api/import-export/import-container) "complex" objects are included in collapsible sections.

I'd also leave out the information about where the sections are positioned, given that the section surrounding it are conditional. Its position in the guide and template should be sufficient.

```suggestion
When the listener callback receives more than the `listener` object and the `listener` object has more than five fields across its objects, to assist with readability, consider documenting the full shape of each object in a separate `## Listener objects` section. Use an H3 (`###`) heading for each object and list its fields in a definition list. Examples of this pattern include [`webRequest.onBeforeSendHeaders`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeSendHeaders). 
```

If the listener parameters are simple enough to document inline in the Parameters subsection, omit this section entirely.

> ## Description section

Reviewer: Why add this? Also, why put the description section towards the end of the page rather than the beginning? As far as I can see, in the current pages, any additional descriptions are included in the introduction or in context with other content.

If there is additional detail to convey beyond what fits in the introductory paragraph — such as when and why the event fires in detail, edge cases, interactions with other events or APIs, or platform-specific behavior — include a `## Description` section placed between the Additional objects section (or addListener syntax section, if Additional objects is omitted) and the Examples section. Omit this section if the introductory paragraph is sufficient.

## Examples section

> Provide one or more examples demonstrating how to listen for the event. Each example must have an H3 (`###`) heading with a concise, descriptive title that says what the example does. See our guide on how to add [code examples](/en-US/docs/MDN/Writing_guidelines/Page_structures/Code_examples) for more information.

Reviewer: ```suggestion
Provide one or more examples demonstrating how to listen for the event. Each example must have an H3 (`###`) heading with a concise, descriptive title that says what the example does. See the guide on how to add [code examples](/en-US/docs/MDN/Writing_guidelines/Page_structures/Code_examples) for more information.
```

> Provide one or more examples demonstrating how to listen for the event. Each example must have an H3 (`###`) heading with a concise, descriptive title that says what the example does. See our guide on how to add [code examples](/en-US/docs/MDN/Writing_guidelines/Page_structures/Code_examples) for more information.

Reviewer: Do you need to include an H3 heading if there's only one example?

Author: It’s consistent with the rest of MDN and helps grow examples and link to them. For example, now there’s only one example, and you’re linking to the generic “Examples” section. But then there are two, three, or four examples. Which are you linking to now?

If you have examples on this page _and_ links to more examples on other pages, add a final H3 heading called "More examples" with the links. If you _only_ have examples on another page and none on this page, add the links directly under the H2 heading without any H3 headings.

After all examples, include the `{{WebExtExamples}}` macro, which generates links to relevant examples from the [webextensions-examples](https://github.com/nicolo-ribaudo/webextensions-examples) repository.

## Browser compatibility section

Include `{{Compat}}`, which is automatically populated from the `browser-compat` key in the front matter.

## See also section

> Include links to reference pages and guides related to the current event. For more guidelines, see the [See also section](/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#see_also_section) in the _Writing style guide_.

Reviewer: ```suggestion
Include links to reference pages and guides related to the event. For more guidelines, see the [See also section](/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#see_also_section) in the _Writing style guide_.
```
